test/test-utils/src/main/java/software/amazon/awssdk/testutils/Waiter.java

```diff
@@ -0,0 +1,131 @@
+        this.thingToTry = thingToTry;
+    }
+
+    /**
+     * Create a waiter that attempts executing the provided function until the condition set with {@link #until(Predicate)} is
+     * met or until it throws an exception. Expected exception types can be ignored with {@link #ignoring(Class[])}.
+     */
+    public static <T> Waiter<T> run(Supplier<T> thingToTry) {
+        return new Waiter<>(thingToTry);
+    }
+
+    /**
+     * Define the condition under which the thing we are trying is complete.
+     *
+     * If this isn't set, it will always be true. ie. if the function call succeeds, we stop waiting.
+     */
+    public Waiter<T> until(Predicate<T> whenToStop) {
+        this.whenToStop = whenToStop;
+        return this;
+    }
+
+    /**
+     * Define the exception types that should be ignored if the thing we are trying throws them.
+     */
+    @SafeVarargs
+    public final Waiter<T> ignoring(Class<? extends Throwable>... whatToIgnore) {
+        this.whatToIgnore = new HashSet<>(Arrays.asList(whatToIgnore));
+        return this;
+    }
+
+    /**
+     * Execute the function, throwing an assertion error if the thing we're trying does not succeed after 30 seconds.
+     */
+    public T orFail() {
+        return orFailAfter(Duration.ofSeconds(30));
+    }
+
+    /**
+     * Execute the function, throwing an assertion error if the thing we're trying does not succeed after the provided duration.
+     */
+    public T orFailAfter(Duration howLongToTry) {
+        Validate.paramNotNull(howLongToTry, "howLongToTry");
+
+        Instant start = Instant.now();
+        int attempt = 0;
+
+        while (Duration.between(start, Instant.now()).compareTo(howLongToTry) < 0) {
+            ++attempt;
+            try {
+                if (attempt > 0) {
+                    wait(attempt);
+                }
+
+                T result = thingToTry.get();
+                if (whenToStop.test(result)) {
+                    return result;
+                }
+                int unsuccessfulAttempt = attempt;
+                log.info(() -> "Attempt " + unsuccessfulAttempt + " failed predicate.");
+            } catch (RuntimeException e) {
+                Throwable t = e instanceof CompletionException ? e.getCause() : e;
+
+                if (whatToIgnore.contains(t.getClass())) {
+                    int unsuccessfulAttempt = attempt;
+                    log.info(() -> "Attempt " + unsuccessfulAttempt +
+                                   " failed with an expected exception (" + t.getClass() + ")");
+                } else {
+                    throw e;
+                }
+            }
+        }
+
+        throw new AssertionError("Condition was not met after " + attempt + " attempts (" +
+                                 Duration.between(start, Instant.now()).getSeconds() + " seconds)");
+    }
+
+    private void wait(int attempt) {
+        try {
+            int howLongToWait = 1000 << attempt;
```

Comment:
I reckon we should have a more aggressive wait initially - what if we started this at 200 ms? As it stands the waits are going to be:
attempt 1 -> 2000 
attempt 2 -> 4000
every other attempt -> 5000

Reply:
Yeah, there's also an overflow risk I've fixed. I've done (250, 500, 1000, 2000, 4000) in the latest.

Reply:
What if we could throw in a guess of how long we think it will take and then grow from there?

If I think it takes about 4 seconds, then start at 4000 and go 4250, 4500, ...)